### PR TITLE
UPC 73 Prevent plugin to have a non-valid key

### PR DESCRIPTION
**Pull request by erichirlemann**

### CODE REVIEW

File: sonar-update-center-common/src/main/java/org/sonar/updatecenter/common/Plugin.java [hunk 1]
```diff
@@ -19,6 +19,7 @@
  */
 package org.sonar.updatecenter.common;
 
+import org.apache.commons.lang.IllegalClassException;
```
Comment (SonarTech) [inline sonar-update-center-common/src/main/java/org/sonar/updatecenter/common/Plugin.java]: MINOR Remove this unused import 'org.apache.commons.lang.IllegalClassException'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-update-center-common/src/main/java/org/sonar/updatecenter/common/Plugin.java [hunk 2]
```diff
@@ -39,10 +40,19 @@
   private boolean supportedBySonarSource = false;
   private List<String> developers;
 
-  public Plugin(String key) {
+  private Plugin(String key) {
     super(key);
   }
 
+  public static Plugin factory(String key) throws IllegalArgumentException {
```
Comment (SonarTech) [inline sonar-update-center-common/src/main/java/org/sonar/updatecenter/common/Plugin.java]: MAJOR Remove the declaration of thrown exception 'java.lang.IllegalArgumentException' which is a runtime exception. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: sonar-update-center-common/src/main/java/org/sonar/updatecenter/common/Plugin.java [hunk 3]
```diff
@@ -39,10 +39,19 @@
   private boolean supportedBySonarSource = false;
   private List<String> developers;
 
-  public Plugin(String key) {
+  private Plugin(String key) {
     super(key);
   }
 
+  public static Plugin factory(String key) {
```
Comment (henryju) [inline sonar-update-center-common/src/main/java/org/sonar/updatecenter/common/Plugin.java]: We usually name such method "create"

**Comment by erichirlemann** [comment]
1/ Use of a factory prevents having a constructor throwing an exception.
The immediate impact on SonarQube is none, since SQ is dependend on an old version 1.13 of Update-Center.
The later impact on SQ is limited, it has no dependency in its source code,. There are dependencies in the unit tests, where a search&replace almost does the job.

2/ During the build process of a plugin, the key is sanitized by [this code|https://github.com/SonarSource/sonar-packaging-maven-plugin/blob/master/src/main/java/org/sonarsource/pluginpackaging/PluginKeyUtils.java#L44], preventing any plugin with bad key to be installed. The check on key validity will rather trigger on user input in .properties file

**Comment by henryju** [comment]
1/ The constructor is private so I still don't see the benefit: the only way to create a new instance will throw an exception. Very minor BTW ;)

2/ ok

So you can go on and merge :)

**Pull request merged**